uint32 data
bool shouldCheckForID
uint32[] agentIDs